File: src/game/RoadPath.ts
import * as THREE from 'three';
import type { TrackRoad } from './formats';

export interface RoadQuery {
  /** Index of the nearest resampled point. */
  index: number;
  /** Arc length from the start of the road to that point, in metres. */
  distance: number;
  /** Signed lateral offset: negative is left of the racing direction. */
  lateral: number;
  point: THREE.Vector3;
  tangent: THREE.Vector3;
  width: number;
}

/**
 * The racing line, resampled to uniform arc length.
 *
 * One object serves three jobs, which is deliberate — they must agree:
 *   - the visible road ribbon is lofted along it,
 *   - the terrain is flattened underneath it,
 *   - the AI follows it and the race director measures progress along it.
 *
 * If these used separate representations they would drift apart and the AI
 * would drive somewhere the road isn't.
 */
export class RoadPath {
  readonly closed: boolean;
  readonly length: number;
  readonly points: THREE.Vector3[] = [];
  readonly tangents: THREE.Vector3[] = [];
  readonly widths: number[] = [];
  readonly banks: number[] = [];
  /** Uniform spacing between resampled points, in metres. */
  readonly step: number;
  readonly shoulder: number;

  private buckets = new Map<number, number[]>();
  private cellSize: number;

  constructor(road: TrackRoad, resolution = 1.5) {
    this.closed = road.closed;
    this.shoulder = road.shoulder;

    const controls = road.points.map((p) => new THREE.Vector3(p.pos[0], p.pos[1], p.pos[2]));
    const curve = new THREE.CatmullRomCurve3(controls, road.closed, 'centripetal', 0.5);

    const approxLength = curve.getLength();
    const count = Math.max(8, Math.round(approxLength / resolution));
    this.step = approxLength / count;
    this.length = approxLength;

    // getSpacedPoints returns count+1 arc-length-uniform samples. For a closed
    // loop the last duplicates the first, so drop it to keep spacing even.
    const spaced = curve.getSpacedPoints(count);
    const sampleCount = road.closed ? count : count + 1;

    for (let i = 0; i < sampleCount; i++) {
      this.points.push(spaced[i].clone());

      // Map the sample back onto control-point space so per-point width and
      // bank overrides interpolate along the same parameterisation.
      const u = i / count;
      const t = curve.getUtoTmapping(u, u * approxLength);
      const span = road.closed ? controls.length : controls.length - 1;
      const raw = t * span;
      const i0 = Math.floor(raw) % controls.length;
      const i1 = (i0 + 1) % controls.length;
      const frac = raw - Math.floor(raw);

      const w0 = road.points[i0].width ?? road.width;
      const w1 = road.points[i1].width ?? road.width;
      this.widths.push(w0 + (w1 - w0) * frac);

      const b0 = road.points[i0].bank ?? 0;
      const b1 = road.points[i1].bank ?? 0;
      this.banks.push(b0 + (b1 - b0) * frac);
    }

    for (let i = 0; i < this.points.length; i++) {
      const next = this.points[this.wrap(i + 1)];
      const prev = this.points[this.wrap(i - 1)];
      this.tangents.push(next.clone().sub(prev).normalize());
    }

    // Cells must be at least as wide as the furthest query we expect to
    // resolve from a 3x3 neighbourhood.
    const maxWidth = Math.max(...this.widths);
    this.cellSize = Math.max(8, maxWidth * 0.5 + road.shoulder + 4);
    this.buildBuckets();
  }

  /** Clamp (open road) or wrap (closed circuit) a sample index. */
  wrap(index: number): number {
    const n = this.points.length;
    if (this.closed) return ((index % n) + n) % n;
    return index < 0 ? 0 : index >= n ? n - 1 : index;
  }

  private key(cx: number, cz: number): number {
    // Pack two signed cell coordinates into one integer key.
    return (cx + 4096) * 8192 + (cz + 4096);
  }

  private buildBuckets(): void {
    for (let i = 0; i < this.points.length; i++) {
      const p = this.points[i];
      const cx = Math.floor(p.x / this.cellSize);
      const cz = Math.floor(p.z / this.cellSize);
      const k = this.key(cx, cz);
      const list = this.buckets.get(k);
      if (list) list.push(i);
      else this.buckets.set(k, [i]);
    }
  }

  /** Arc length at a sample index. */
  distanceAt(index: number): number {
    return this.wrap(index) * this.step;
  }

  pointAt(index: number): THREE.Vector3 {
    return this.points[this.wrap(index)];
  }

  tangentAt(index: number): THREE.Vector3 {
    return this.tangents[this.wrap(index)];
  }

  widthAt(index: number): number {
    return this.widths[this.wrap(index)];
  }

  /**
   * Nearest point on the road to a world position.
   *
   * Uses a uniform bucket grid; falls back to a strided brute-force sweep for
   * queries far off the course, which happens when a truck gets launched.
   */
  closestTo(x: number, z: number): RoadQuery {
    const cx = Math.floor(x / this.cellSize);
    const cz = Math.floor(z / this.cellSize);

    let best = -1;
    let bestDistanceSq = Infinity;

    for (let dx = -1; dx <= 1; dx++) {
      for (let dz = -1; dz <= 1; dz++) {
        const list = this.buckets.get(this.key(cx + dx, cz + dz));
        if (!list) continue;
        for (const i of list) {
          const p = this.points[i];
          const d = (p.x - x) ** 2 + (p.z - z) ** 2;
          if (d < bestDistanceSq) {
            bestDistanceSq = d;
            best = i;
          }
        }
      }
    }

    if (best < 0) {
      const stride = Math.max(1, Math.floor(this.points.length / 256));
      for (let i = 0; i < this.points.length; i += stride) {
        const p = this.points[i];
        const d = (p.x - x) ** 2 + (p.z - z) ** 2;
        if (d < bestDistanceSq) {
          bestDistanceSq = d;
          best = i;
        }
      }
      // Refine around the coarse hit so the result is still accurate.
      const coarse = best;
      for (let i = coarse - stride; i <= coarse + stride; i++) {
        const p = this.points[this.wrap(i)];
        const d = (p.x - x) ** 2 + (p.z - z) ** 2;
        if (d < bestDistanceSq) {
          bestDistanceSq = d;
          best = this.wrap(i);
        }
      }
    }

    const point = this.points[best];
    const tangent = this.tangents[best];
    // Sign the offset by which side of the tangent the query falls on.
    const toQueryX = x - point.x;
    const toQueryZ = z - point.z;
    const cross = tangent.z * toQueryX - tangent.x * toQueryZ;
    const lateral = Math.sign(cross) * Math.sqrt(bestDistanceSq);

    return {
      index: best,
      distance: best * this.step,
      lateral,
      point,
      tangent,
      width: this.widths[best],
    };
  }

  /**
   * Heading in radians at a sample, measured so that it can be handed
   * straight to a Y-axis rotation.
   */
  headingAt(index: number): number {
    const t = this.tangentAt(index);
    return Math.atan2(t.x, t.z);
  }

  /**
   * Signed curvature over a lookahead window, used by the AI to decide how
   * hard to brake for what is coming.
   */
  curvatureAt(index: number, lookaheadSamples: number): number {
    const a = this.tangentAt(index);
    const b = this.tangentAt(index + lookaheadSamples);
    const cross = a.z * b.x - a.x * b.z;
    const dot = Math.max(-1, Math.min(1, a.dot(b)));
    return Math.sign(cross) * Math.acos(dot);
  }
}

File: unity/MonsterTruckMania/Assets/MonsterTruckMania/Runtime/Generation/RoadPath.cs
// SPDX-License-Identifier: MIT
using System;
using System.Collections.Generic;

namespace MonsterTruckMania.Generation
{
    /// <summary>
    /// The road centreline: a Catmull-Rom curve resampled at even arc length,
    /// with a bucket grid for "how far is this point from the road".
    /// </summary>
    /// <remarks>
    /// One source of truth for the racing line. The visible road ribbon, the
    /// terrain flattening under it, the prop scatter exclusion and the
    /// out-of-bounds test all read from this same resampled polyline. Deriving
    /// them separately is how you end up with a road that is not quite where
    /// the terrain thinks it is.
    /// </remarks>
    public sealed class RoadPath
    {
        private readonly Dictionary<long, List<int>> _buckets = new Dictionary<long, List<int>>();
        private readonly double _cell;

        public IReadOnlyList<Vec3> Points { get; }
        /// <summary>Spacing between samples, in metres.</summary>
        public double Step { get; }
        /// <summary>Total curve length, in metres.</summary>
        public double Length { get; }
        public double Width { get; }
        public double Shoulder { get; }
        public bool Closed { get; }

        public RoadPath(IReadOnlyList<Vec3> control, bool closed, double width, double shoulder,
                        double resolution = 1.5)
        {
            Width = width;
            Shoulder = shoulder;
            Closed = closed;

            if (control.Count < 2)
            {
                Points = new List<Vec3>(control);
                Step = 0.0;
                Length = 0.0;
                _cell = Math.Max(8.0, width * 0.5 + shoulder + 4.0);
                return;
            }

            var curve = new CatmullRomCurve(control, closed);
            Length = curve.GetLength();
            int count = Math.Max(8, (int)Math.Round(Length / resolution, MidpointRounding.AwayFromZero));
            List<Vec3> spaced = curve.GetSpacedPoints(count);

            // GetSpacedPoints returns divisions+1 samples, and on a loop the
            // last repeats the first. Keeping it leaves a zero-length segment
            // right on the start line, which shows up as a pinched quad in the
            // road mesh and a divide-by-zero in the tangent there.
            int sampleCount = closed ? count : count + 1;
            var points = new List<Vec3>(sampleCount);
            for (int i = 0; i < sampleCount; i++) points.Add(spaced[i]);
            Points = points;
            Step = Length / count;

            _cell = Math.Max(8.0, width * 0.5 + shoulder + 4.0);
            BuildBuckets();
        }

        /// <summary>
        /// Build from samples that are already spaced, skipping the spline.
        /// </summary>
        /// <remarks>
        /// For a road that came from data rather than from control points —
        /// a track authored elsewhere, or a baked path — where re-splining
        /// would move it slightly. Also what the terrain tests use, so a
        /// straight road can be stated exactly rather than approximated by
        /// control points that happen to produce one.
        /// </remarks>
        public static RoadPath FromSamples(IReadOnlyList<Vec3> samples, bool closed,
                                           double width, double shoulder, double step)
        {
            return new RoadPath(samples, closed, width, shoulder, step, true);
        }

        private RoadPath(IReadOnlyList<Vec3> samples, bool closed, double width, double shoulder,
                         double step, bool preSampled)
        {
            _ = preSampled;
            Width = width;
            Shoulder = shoulder;
            Closed = closed;
            Points = new List<Vec3>(samples);
            Step = step;

            double length = 0.0;
            for (int i = 0; i < Points.Count - 1; i++) length += Vec3.Distance(Points[i], Points[i + 1]);
            Length = length;

            _cell = Math.Max(8.0, width * 0.5 + shoulder + 4.0);
            BuildBuckets();
        }

        private void BuildBuckets()
        {
            for (int i = 0; i < Points.Count; i++)
            {
                long key = Key((int)Math.Floor(Points[i].X / _cell), (int)Math.Floor(Points[i].Z / _cell));
                if (!_buckets.TryGetValue(key, out var list))
                {
                    list = new List<int>();
                    _buckets[key] = list;
                }
                list.Add(i);
            }
        }

        private static long Key(int cx, int cz) => ((long)cx << 32) ^ (uint)cz;

        /// <summary>
        /// How far from the road <see cref="Closest"/> is guaranteed to find
        /// the true nearest sample.
        /// </summary>
        /// <remarks>
        /// The lookup searches the nine buckets around the query, so anything
        /// within one cell is certain to be found and anything beyond it is
        /// best-effort. Sized so the whole carved corridor — road plus
        /// shoulder — is comfortably inside the guarantee, since that is what
        /// the terrain carve queries.
        /// </remarks>
        public double SearchRadius => _cell;

        /// <summary>
        /// Nearest sample to a world XZ position, and the distance to it.
        /// </summary>
        /// <remarks>
        /// Only the nine buckets around the query are searched, which is what
        /// makes carving a 256x256 grid against a 1500m course tolerable —
        /// the naive version is O(grid * samples) and takes seconds.
        /// Returns index -1 when nothing is near, so callers can skip.
        /// </remarks>
        public int Closest(double x, double z, out double lateral)
        {
            int cx = (int)Math.Floor(x / _cell);
            int cz = (int)Math.Floor(z / _cell);
            int best = -1;
            double bestSq = double.PositiveInfinity;

            for (int dx = -1; dx <= 1; dx++)
            {
                for (int dz = -1; dz <= 1; dz++)
                {
                    if (!_buckets.TryGetValue(Key(cx + dx, cz + dz), out var list)) continue;
                    for (int k = 0; k < list.Count; k++)
                    {
                        Vec3 p = Points[list[k]];
                        double d = (p.X - x) * (p.X - x) + (p.Z - z) * (p.Z - z);
                        if (d < bestSq)
                        {
                            bestSq = d;
                            best = list[k];
                        }
                    }
                }
            }

            lateral = best < 0 ? double.PositiveInfinity : Math.Sqrt(bestSq);
            return best;
        }

        /// <summary>Direction of travel at a sample, normalised, XZ only.</summary>
        public Vec3 TangentAt(int index)
        {
            int count = Points.Count;
            if (count < 2) return new Vec3(0, 0, 1);
            int next = Closed ? (index + 1) % count : Math.Min(index + 1, count - 1);
            int prev = Closed ? (index - 1 + count) % count : Math.Max(index - 1, 0);
            Vec3 d = Points[next] - Points[prev];
            return new Vec3(d.X, 0, d.Z).Normalized;
        }

        /// <summary>Right-hand normal at a sample, in the XZ plane.</summary>
        /// <remarks>
        /// This is cross(up, tangent), not cross(tangent, up) — the latter
        /// points left. Unity's frame is left-handed, so facing +Z your right
        /// is +X, and facing +X your right is -Z; both fall out of the form
        /// below. Everything that offsets from the centreline (grid columns,
        /// AI lane offsets, the sign of <see cref="RoadQuery.Lateral"/>) reads
        /// this, so a flip here mirrors the whole course.
        /// </remarks>
        public Vec3 RightAt(int index)
        {
            Vec3 t = TangentAt(index);
            return new Vec3(t.Z, 0, -t.X);
        }

        /// <summary>Clamp (open road) or wrap (closed circuit) a sample index.</summary>
        /// <remarks>
        /// Every lookahead in the game adds samples to an index and expects
        /// the result to stay valid, so the wrap lives here rather than at
        /// each call site. An open road clamps instead: running off the end
        /// of a point-to-point course must not teleport the AI to the start.
        /// </remarks>
        public int Wrap(int index)
        {
            int n = Points.Count;
            if (n == 0) return 0;
            if (Closed) return ((index % n) + n) % n;
            return index < 0 ? 0 : index >= n ? n - 1 : index;
        }

        /// <summary>Sample position, index wrapped or clamped.</summary>
        public Vec3 PointAt(int index) => Points.Count == 0 ? new Vec3(0, 0, 0) : Points[Wrap(index)];

        /// <summary>Arc length from the start of the road to a sample.</summary>
        public double DistanceAt(int index) => Wrap(index) * Step;

        /// <summary>
        /// Heading at a sample, as a rotation about +Y.
        /// </summary>
        /// <remarks>
        /// Measured as atan2(x, z) rather than the usual atan2(z, x) so it can
        /// be handed straight to a Y-axis rotation with a +Z-forward model —
        /// which is the convention every truck, prop and spawn in this project
        /// uses. Converting at each call site is how half of them end up 90
        /// degrees out.
        /// </remarks>
        public double HeadingAt(int index)
        {
            Vec3 t = TangentAt(index);
            return Math.Atan2(t.X, t.Z);
        }

        /// <summary>
        /// Signed change of direction over a lookahead window, in radians.
        /// </summary>
        /// <remarks>
        /// This is what the AI brakes on: not the instantaneous curvature,
        /// which is noisy on a resampled polyline, but how much the road has
        /// turned by the time it gets there. Positive is a right-hand bend.
        /// </remarks>
        public double CurvatureAt(int index, int lookaheadSamples)
        {
            Vec3 a = TangentAt(Wrap(index));
            Vec3 b = TangentAt(Wrap(index + lookaheadSamples));
            double cross = a.Z * b.X - a.X * b.Z;
            double dot = MathUtil.Clamp(a.X * b.X + a.Z * b.Z, -1.0, 1.0);
            return Math.Sign(cross) * Math.Acos(dot);
        }

        /// <summary>
        /// Everything a driver or the race director needs about where a world
        /// position sits relative to the road.
        /// </summary>
        /// <remarks>
        /// <see cref="Closest"/> answers "how far from the road is this",
        /// which is all the terrain carve needs, and it gives up outside the
        /// bucket neighbourhood. A truck that has been launched off a jump —
        /// or shoved into the scenery — is routinely outside it, and the AI
        /// and the lap counter cannot give up on those frames. So this falls
        /// back to a strided sweep of the whole path and then refines around
        /// the coarse hit, which is exact to the sample and still cheap.
        /// </remarks>
        public RoadQuery Query(double x, double z)
        {
            if (Points.Count == 0) return default;

            int best = Closest(x, z, out double _);
            if (best < 0) best = SweepForClosest(x, z);

            Vec3 point = Points[best];
            Vec3 tangent = TangentAt(best);
            double dx = x - point.X;
            double dz = z - point.Z;
            double distance = Math.Sqrt(dx * dx + dz * dz);

            // Which side of the road: the sign of the cross product of the
            // direction of travel with the offset. Positive is the right-hand
            // side going forwards, which is what the AI's lane offsets and the
            // reverse-out manoeuvre both assume.
            double cross = tangent.Z * dx - tangent.X * dz;

            return new RoadQuery(best, DistanceAt(best), Math.Sign(cross) * distance, point, tangent, Width);
        }

        private int SweepForClosest(double x, double z)
        {
            int count = Points.Count;
            int stride = Math.Max(1, count / 256);
            int best = 0;
            double bestSq = double.PositiveInfinity;

            for (int i = 0; i < count; i += stride)
            {
                double d = SqrPlanarDistance(i, x, z);
                if (d < bestSq) { bestSq = d; best = i; }
            }

            int coarse = best;
            for (int i = coarse - stride; i <= coarse + stride; i++)
            {
                int w = Wrap(i);
                double d = SqrPlanarDistance(w, x, z);
                if (d < bestSq) { bestSq = d; best = w; }
            }
            return best;
        }

        private double SqrPlanarDistance(int index, double x, double z)
        {
            Vec3 p = Points[index];
            return (p.X - x) * (p.X - x) + (p.Z - z) * (p.Z - z);
        }
    }

    /// <summary>
    /// Where a world position sits relative to the road.
    /// </summary>
    public readonly struct RoadQuery
    {
        /// <summary>Index of the nearest resampled sample.</summary>
        public readonly int Index;
        /// <summary>Arc length from the start of the road to that sample.</summary>
        public readonly double Distance;
        /// <summary>Signed offset from the centreline; negative is left of travel.</summary>
        public readonly double Lateral;
        public readonly Vec3 Point;
        public readonly Vec3 Tangent;
        public readonly double Width;

        public RoadQuery(int index, double distance, double lateral, Vec3 point, Vec3 tangent, double width)
        {
            Index = index;
            Distance = distance;
            Lateral = lateral;
            Point = point;
            Tangent = tangent;
            Width = width;
        }

        /// <summary>How far outside the road plus its shoulder this is, in metres.</summary>
        public double DistanceOutside(double shoulder) => Math.Abs(Lateral) - (Width * 0.5 + shoulder);
    }
}
